src/azure-cli-core/azure/cli/core/_identity.py

```diff
@@ -0,0 +1,268 @@
+
+from knack.util import CLIError
+from knack.log import get_logger
+
+from azure.identity import (
+    AuthProfile,
+    InteractiveBrowserCredential,
+    DeviceCodeCredential,
+    UsernamePasswordCredential,
+    ClientSecretCredential,
+    CertificateCredential,
+    ManagedIdentityCredential
+)
+
+_CLIENT_ID = '04b07795-8ddb-461a-bbee-02f9e1bf7b46'
+logger = get_logger(__name__)
+
+_SERVICE_PRINCIPAL_ID = 'servicePrincipalId'
+_SERVICE_PRINCIPAL_TENANT = 'servicePrincipalTenant'
+_ACCESS_TOKEN = 'accessToken'
+_SERVICE_PRINCIPAL_CERT_FILE = 'certificateFile'
+_SERVICE_PRINCIPAL_CERT_THUMBPRINT = 'thumbprint'
+
+
+def _load_tokens_from_file(file_path):
+    if os.path.isfile(file_path):
+        try:
+            return get_file_json(file_path, throw_on_empty=False) or []
+        except (CLIError, ValueError) as ex:
+            raise CLIError("Failed to load token files. If you have a repro, please log an issue at "
+                           "https://github.com/Azure/azure-cli/issues. At the same time, you can clean "
+                           "up by running 'az account clear' and then 'az login'. (Inner Error: {})".format(ex))
+    return []
+
+
+def _delete_file(file_path):
+    try:
+        os.remove(file_path)
+    except FileNotFoundError:
+        pass
+
+
+class Identity:
+    """Class to interact with Azure Identity.
+    """
+    def __init__(self, authority, tenant_id):
+        self.authority = authority
+        self.tenant_id = tenant_id
+
+    def login_with_interactive_browser(self):
+        # Use InteractiveBrowserCredential
+        if self.tenant_id:
+            credential, auth_profile = InteractiveBrowserCredential.authenticate(
+                client_id=_CLIENT_ID,
+                tenant_id=self.tenant_id
+            )
+        else:
+            credential, auth_profile = InteractiveBrowserCredential.authenticate(
+                client_id=_CLIENT_ID,
+                silent_auth_only=True
+            )
+        return credential, auth_profile
+
+    def login_with_device_code(self):
+        # Use DeviceCodeCredential
+        message = 'To sign in, use a web browser to open the page {} and enter the code {} to authenticate.'
+        prompt_callback=lambda verification_uri, user_code, expires_on: \
+            logger.warning(message.format(verification_uri, user_code))
+        if self.tenant_id:
+            cred, auth_profile = DeviceCodeCredential.authenticate(client_id=_CLIENT_ID,
+                                                                   tenant_id=self.tenant_id,
+                                                                   prompt_callback=prompt_callback)
+        else:
+            cred, auth_profile = DeviceCodeCredential.authenticate(client_id=_CLIENT_ID,
+                                                                   prompt_callback=prompt_callback)
+        return cred, auth_profile
+
+    def login_with_username_password(self, username, password):
```

Comment:
>username [](start = 43, length = 8)

add parameter None check for all these functions

Reply:
They are already checked in upper level:

https://github.com/Azure/azure-cli/blob/d9b5c6386335b5d451a4fd85f13b45b00d816056/src/azure-cli/azure/cli/command_modules/profile/custom.py#L132-L137

Reply:
And here

https://github.com/Azure/azure-cli/blob/d9b5c6386335b5d451a4fd85f13b45b00d816056/src/azure-cli/azure/cli/command_modules/profile/custom.py#L111-L119